packaging/linux/run_keybase

```diff
@@ -111,13 +132,22 @@ start_systemd() {
       QT4_IM_MODULE
       XMODIFIERS
   )
-  cat /dev/null > "$runtime_dir/keybase.gui.env"
-  for varname in "${forwarded_env_vars[@]}" ; do
-      # Include set-but-empty variables but not unset variables.
-      if [ -n "${!varname+x}" ] ; then
-          echo "$varname=${!varname}" >> "$runtime_dir/keybase.gui.env"
-      fi
-  done
+  forward_vars "$runtime_dir/keybase.gui.env" forwarded_gui_env_vars[@]
+
+  mkdir -p "$mountdir"
+  keybase-mount-helper -user-mount "$mountdir"
+
+  forwarded_kbfs_env_vars=(
+      # Needed to determine the mountdir inside the service.
+      KEYBASE_MOUNTDIR
+  )
+  forward_vars "$runtime_dir/kbfs.env" forwarded_kbfs_env_vars[@]
+
+  forwarded_service_env_vars=(
+      # Needed to determine the mountdir inside the service.
+      KEYBASE_MOUNTDIR
+  )
```

Comment:
Passing this parameter as an env var feels a little bit like going against the grain to me. Would it be simpler to have a symlink at say `~/.config/keybase/mountdir` and to have `kbfs.service` just hardcode that symlink?

Reply:
Systemd unit files support passing an "instance name" after an `@` sign in the name of the service. I wonder if we could set this up so that we call `systemctl --user start kbfs@/foo/bar/baz.service`. I haven't yet tried it though. Some of the [documentation is here](https://www.freedesktop.org/software/systemd/man/systemd.unit.html), if you search for "@" and "instance name".

Reply:
Actually `$KEYBASE_MOUNTDIR` is needed as an environment variable in the `libkb.Env.GetMountDir()` function.  I can expand the comment about that. There's a KBFS ticket to get rid of this needed.

I don't understand your proposal about the instance name, not sure what that has to do with passing the mountpoint to the service.

Note that you commented on the server/`keybase` list, not the kbfs/`kbfsfuse` list.  Maybe you meant to be asking about kbfs?  Still, this seems way simpler than adding another file to me, but I admit I'm totally unfamiliar with systemd conventions.  I think it's probably still best to have it in the env though, in case we ever call `libkb.Env.GetMountDir()` inside KBFS.  For example, maybe we want to have a default mountpoint, instead of making it a required argument.

Reply:
What I'm thinking of is an interesting aspect of systemd, where they support having more than once instance of a given service, with each instance distinguished by some parameter. For example on my system, I have `user@1000.service` and `user@1001.service`. Both of these refer to the same service file, `/usr/lib/systemd/system/user@.service`, which uses the `%i` "specifier" in its unit file to refer to "whatever it is that comes between `@` and `.service`". That could be a more with-the-grain way to pass a parameter down to kbfsfuse, especially if we ever wanted to allow more than one of them to run at the same time for the same user.

Reply:
I'm thinking about cases where I might start and stop `kbfs.service` manually, with something like `systemctl --user start kbfs.service`. It looks like with the current PR, that command will appear to work if I've already run `run_keybase` at some point since boot, because the environment fill will be sticking around in my runtime dir. But it won't work if I haven't run `run_keybase` to create that file. That means that trying something like `systemctl --user enable kbfs.service` to get it to autostart in a headless environment, probably isn't going to work without some extra hacks.

Reply:
The GUI service's whitelist isn't necessary on newer distro versions, because all those env vars are available by default now, and so start and stop work fine by hand. But here we're adding some required parameters that can only be passed through an environment file on all systems.

Anyway, I don't think I'm doing a good job of giving you concrete suggestions :p Let me know if you want to hop on a hangout this afternoon maybe?

Reply:
Thanks @oconnor663, I can experiment with passing it as a specifier.  But would that make sense for the `keybase` service as well, which needs to set it in the environment?  It doesn't really seem like the right thing to parameterize the service on.

So you're also suggesting moving the mountpoint creation and mount-helper invocation in the kbfs service definition for systemd?  Seems fine to me as long as we can still print the output of it to the user (so they know whether they got `/keybase` or not).  I didn't realize we were trying to support non-run_keybase invocations.

Reply:
Oh that's a good point, I hadn't thought about the helper. Maybe the way I'm imagining all this doesn't make a ton of sense. I'll think about it while I look at the latest changes below.